pkg/virt-controller/watch/vmi.go

```diff
@@ -1836,57 +1986,54 @@ func (c *VMIController) getActiveAndOldAttachmentPods(readyHotplugVolumes []*vir
 			currentPod = attachmentPod
 		}
 	}
+	sort.Slice(oldPods, func(i, j int) bool {
+		return oldPods[i].CreationTimestamp.Time.After(oldPods[j].CreationTimestamp.Time)
+	})
+	if currentPod == nil && len(oldPods) > 0 {
+		currentPod = oldPods[0]
+		oldPods = oldPods[1:]
+	}
 	return currentPod, oldPods
 }
 
 func (c *VMIController) handleHotplugVolumes(hotplugVolumes []*virtv1.Volume, hotplugAttachmentPods []*k8sv1.Pod, vmi *virtv1.VirtualMachineInstance, virtLauncherPod *k8sv1.Pod, dataVolumes []*cdiv1.DataVolume) syncError {
 	logger := log.Log.Object(vmi)
+	key, err := controller.KeyFunc(vmi)
+	if err != nil {
+		logger.Object(vmi).Reason(err).Error("failed to extract key from virtualmachine.")
+		return &syncErrorImpl{fmt.Errorf("failed to extract key from virtualmachine. %v", err), "FailedToExtractKeyFromVirtualMachine"}
+	}
+
+	c.attachmentPodQueue.Add(key)
+	return nil
+}
 
+func (c *VMIController) getReadyHotplugVolumes(hotplugVolumes []*virtv1.Volume, vmi *virtv1.VirtualMachineInstance, dataVolumes []*cdiv1.DataVolume, virtLauncherPod *k8sv1.Pod) ([]*virtv1.Volume, error) {
+	logger := log.Log.Object(vmi)
 	readyHotplugVolumes := make([]*virtv1.Volume, 0)
-	// Find all ready volumes
+
 	for _, volume := range hotplugVolumes {
 		var err error
 		ready, wffc, err := storagetypes.VolumeReadyToAttachToNode(vmi.Namespace, *volume, dataVolumes, c.dataVolumeInformer, c.pvcInformer)
 		if err != nil {
-			return &syncErrorImpl{fmt.Errorf("Error determining volume status %v", err), PVCNotReadyReason}
+			return nil, &syncErrorImpl{fmt.Errorf("Error determining volume status %v", err), PVCNotReadyReason}
 		}
 		if wffc {
-			// Volume in WaitForFirstConsumer, it has not been populated by CDI yet. create a dummy pod
-			logger.V(1).Infof("Volume %s/%s is in WaitForFistConsumer, triggering population", vmi.Namespace, volume.Name)
-			syncError := c.triggerHotplugPopulation(volume, vmi, virtLauncherPod)
-			if syncError != nil {
-				return syncError
+
+			logger.V(3).Infof("Volume %s/%s is in WaitForFistConsumer, triggering population", vmi.Namespace, volume.Name)
+			err := c.triggerHotplugPopulation(volume, vmi, virtLauncherPod)
+			if err != nil {
+				return nil, err
 			}
 			continue
 		}
 		if !ready {
-			// Volume not ready, skip until it is.
 			logger.V(3).Infof("Skipping hotplugged volume: %s, not ready", volume.Name)
 			continue
 		}
 		readyHotplugVolumes = append(readyHotplugVolumes, volume)
 	}
-	// Determine if the ready volumes have changed compared to the current pod
-	currentPod, oldPods := c.getActiveAndOldAttachmentPods(readyHotplugVolumes, hotplugAttachmentPods)
-
-	if currentPod == nil && len(readyHotplugVolumes) > 0 {
-		// ready volumes have changed
-		// Create new attachment pod that holds all the ready volumes
```

Comment:
This pr is more involved than I thought it would be. What's the reasoning behind using another queue when it seems like we could use the existing vmi queue to express this.

For example, before the hotplug pod is created, we could check if enough time has past since the last pod was created, and if not do something like `c.Queue.AddAfter(vmiKey, time.Duration(timeLeft)*time.Second) `. That will force the vmi to be reconciled again at some point after the specified time, which will ensure we eventually create the new hotplug pod.

Reply:
So initially I thought I wanted a different type of queue, but it turns out the original queue actually had all the properties I wanted. But by then I had split out all the logic and I kind of like it. Keeps a clear split between the logic for the VMI, and the logic for managing the attachment pods. 

From a technical perspective I could remove the extra queue and re-use the same one, but I like having the split. It would not be a lot of work to revert it though.

Reply:
Having multiple queues in the same controller that both work in parallel on the same object key (vmi) is not a pattern that is consistent with how we build controllers. There are good reasons for this.

Multiple queues break assumptions we have around concurrency within the controller. With a single queue, we are guaranteed that a VMI is only being operated on by a single thread at a single point in time, not multiple concurrent threads. The fact that both queues share the same pod expectations logic is also strange. It shouldn't be possible for two threads both acting on a VMI to be modifying expectation logic for the same key, but that's what is occurring here.

It's important that we never need to call into question the role that concurrency may play when processing a VMI. A single queue provides us that guarantee.

This pr shouldn't need new queuing logic to begin with. We should be able to look at the hotplug volume's creation timestamp compared with the current time, and delay processing the vmi's new hotplug volume using `c.Queue.AddAfter(vmiKey, time.Duration(timeLeft)*time.Second) `.  I wouldn't think this would involve more than maybe 10 or less lines of additional code. i could be wrong though

```
delaySeconds := <calculation of how many seconds to wait before creating new hotplug volume>
diffSeconds := now-podCreationTimestamp
remainingSeconds := 0
if delaySeconds > diffSeconds {
    remainingSeconds = delaySeconds-diffSeconds
    c.Queue.AddAfter(vmiKey, time.Duration(remainingSeconds)*time.Second)
    // add after ensures the queue is popped for this key and we make it back to this point in the future
    return nil
}
<create new hotplug attachment>

```

Reply:
Okay I reworked the PR to not use the extra queue. I have verified it is still doing the exact same thing as with the extra queue.